File: IOS/Sources/PeerOffers.swift
import Foundation

struct PeerOfferFile: Identifiable, Codable, Hashable {
    let id: String
    let name: String
    let size: Int64
    let mime: String
}

struct PeerOfferWire: Codable {
    let protocolName: String
    let offerId: String
    let senderDeviceId: String
    let senderAlias: String
    let senderPlatform: String
    let peerPort: Int
    let token: String
    let verificationCode: String
    let expiresAt: Int64
    let files: [PeerOfferFile]

    enum CodingKeys: String, CodingKey {
        case protocolName = "protocol"
        case offerId, senderDeviceId, senderAlias, senderPlatform, peerPort, token, verificationCode, expiresAt, files
    }
}

struct IncomingPeerOffer: Identifiable, Hashable {
    var id: String { offerId }
    let offerId: String
    let senderDeviceId: String
    let senderAlias: String
    let senderPlatform: String
    let host: String
    let peerPort: Int
    let token: String
    let verificationCode: String
    let expiresAt: Int64
    let files: [PeerOfferFile]
}

struct PeerOfferReceiveResult {
    let urls: [URL]
    let directory: URL?
    let failures: [String]
}

private struct OutgoingPeerGrant {
    let token: String
    let expiresAt: Int64
    var fileIds: Set<String>
}

final class PeerOfferStore {
    static let shared = PeerOfferStore()

    private let lock = NSLock()
    private var incomingById: [String: IncomingPeerOffer] = [:]
    private var outgoingById: [String: OutgoingPeerGrant] = [:]
    private let offerTTL: Int64 = 5 * 60 * 1000
    private let maximumClockWindow: Int64 = 10 * 60 * 1000
    private let maximumIncomingOffers = 20
    private let maximumFilesPerOffer = 128

    private init() {}

    func prepareOutgoing(files: [SharedFile]) throws -> PeerOfferWire {
        guard !files.isEmpty else { throw offerError("No pending files to offer") }
        guard files.count <= maximumFilesPerOffer else { throw offerError("Too many files in one offer") }

        let now = nowMilliseconds()
        let offerId = randomHex(bytes: 16)
        let token = randomHex(bytes: 32)
        let code = String(format: "%06d", Int.random(in: 0..<1_000_000))
        let expiresAt = now + offerTTL
        let manifest = files.map { PeerOfferFile(id: $0.id, name: $0.name, size: $0.size, mime: $0.mime) }
        let wire = PeerOfferWire(
            protocolName: PeerProtocol.versionV2,
            offerId: offerId,
            senderDeviceId: PeerIdentity.deviceId(),
            senderAlias: PeerIdentity.alias(),
            senderPlatform: "iOS",
            peerPort: PeerProtocol.peerPort,
            token: token,
            verificationCode: code,
            expiresAt: expiresAt,
            files: manifest
        )

        lock.lock()
        cleanupLocked(now: now)
        outgoingById[offerId] = OutgoingPeerGrant(token: token, expiresAt: expiresAt, fileIds: Set(manifest.map(\.id)))
        lock.unlock()
        return wire
    }

    func cancelOutgoing(_ offerId: String) {
        lock.lock()
        outgoingById.removeValue(forKey: offerId)
        lock.unlock()
    }

    func receiveIncoming(remoteHost: String, data: Data) -> IncomingPeerOffer? {
        guard let wire = try? JSONDecoder().decode(PeerOfferWire.self, from: data) else { return nil }
        let now = nowMilliseconds()
        guard wire.protocolName == PeerProtocol.versionV2 else { return nil }
        guard isHex(wire.offerId, length: 32), isHex(wire.token, length: 64) else { return nil }
        guard !wire.senderDeviceId.isEmpty, wire.senderDeviceId != PeerIdentity.deviceId() else { return nil }
        guard wire.verificationCode.range(of: "^[0-9]{6}$", options: .regularExpression) != nil else { return nil }
        guard wire.expiresAt > now, wire.expiresAt <= now + maximumClockWindow else { return nil }
        guard wire.peerPort > 0, wire.peerPort <= 65_535, !remoteHost.isEmpty else { return nil }
        guard !wire.files.isEmpty, wire.files.count <= maximumFilesPerOffer else { return nil }

        let files = wire.files.filter {
            !$0.id.isEmpty && $0.id.count <= 64 && $0.id.allSatisfy(\.isLetterOrNumber) && !$0.name.isEmpty && $0.name.count <= 255
        }
        guard !files.isEmpty else { return nil }

        let offer = IncomingPeerOffer(
            offerId: wire.offerId.lowercased(),
            senderDeviceId: wire.senderDeviceId,
            senderAlias: String(wire.senderAlias.prefix(120)).isEmpty ? remoteHost : String(wire.senderAlias.prefix(120)),
            senderPlatform: String(wire.senderPlatform.prefix(40)).isEmpty ? "Unknown" : String(wire.senderPlatform.prefix(40)),
            host: remoteHost,
            peerPort: wire.peerPort,
            token: wire.token.lowercased(),
            verificationCode: wire.verificationCode,
            expiresAt: wire.expiresAt,
            files: files
        )

        lock.lock()
        cleanupLocked(now: now)
        guard incomingById[offer.offerId] != nil || incomingById.count < maximumIncomingOffers else {
            lock.unlock()
            return nil
        }
        incomingById[offer.offerId] = offer
        lock.unlock()
        return offer
    }

    func incoming() -> [IncomingPeerOffer] {
        lock.lock()
        cleanupLocked(now: nowMilliseconds())
        let values = incomingById.values.sorted {
            $0.senderAlias.localizedCaseInsensitiveCompare($1.senderAlias) == .orderedAscending
        }
        lock.unlock()
        return values
    }

    func decline(_ offerId: String) {
        lock.lock()
        incomingById.removeValue(forKey: offerId)
        lock.unlock()
    }

    func markIncomingFileComplete(offerId: String, fileId: String) {
        lock.lock()
        if let offer = incomingById[offerId] {
            let remaining = offer.files.filter { $0.id != fileId }
            if remaining.isEmpty {
                incomingById.removeValue(forKey: offerId)
            } else {
                incomingById[offerId] = IncomingPeerOffer(
                    offerId: offer.offerId,
                    senderDeviceId: offer.senderDeviceId,
                    senderAlias: offer.senderAlias,
                    senderPlatform: offer.senderPlatform,
                    host: offer.host,
                    peerPort: offer.peerPort,
                    token: offer.token,
                    verificationCode: offer.verificationCode,
                    expiresAt: offer.expiresAt,
                    files: remaining
                )
            }
        }
        lock.unlock()
    }

    func authorizeOutgoing(offerId: String, fileId: String, authorization: String?) -> Bool {
        let now = nowMilliseconds()
        lock.lock()
        cleanupLocked(now: now)
        guard let grant = outgoingById[offerId], grant.fileIds.contains(fileId) else {
            lock.unlock()
            return false
        }
        lock.unlock()

        guard let authorization else { return false }
        let components = authorization.split(separator: " ", maxSplits: 1)
        guard components.count == 2, components[0].lowercased() == "bearer" else { return false }
        return String(components[1]).lowercased() == grant.token.lowercased()
    }

    func completeOutgoingFile(offerId: String, fileId: String) {
        lock.lock()
        if var grant = outgoingById[offerId] {
            grant.fileIds.remove(fileId)
            if grant.fileIds.isEmpty { outgoingById.removeValue(forKey: offerId) }
            else { outgoingById[offerId] = grant }
        }
        lock.unlock()
    }

    private func cleanupLocked(now: Int64) {
        incomingById = incomingById.filter { $0.value.expiresAt > now }
        outgoingById = outgoingById.filter { $0.value.expiresAt > now && !$0.value.fileIds.isEmpty }
    }

    private func nowMilliseconds() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }

    private func randomHex(bytes: Int) -> String {
        (0..<bytes).map { _ in String(format: "%02x", UInt8.random(in: UInt8.min...UInt8.max)) }.joined()
    }

    private func isHex(_ value: String, length: Int) -> Bool {
        value.count == length && value.allSatisfy { $0.isHexDigit }
    }

    private func offerError(_ message: String) -> NSError {
        NSError(domain: "FTPortal.PeerOffer", code: 1, userInfo: [NSLocalizedDescriptionKey: message])
    }
}

enum PeerOfferClient {
    static func send(lobby: PeerLobby, files: [SharedFile]) async throws -> PeerOfferWire {
        guard lobby.supportsOffers else {
            throw NSError(domain: "FTPortal.PeerOffer", code: 2, userInfo: [NSLocalizedDescriptionKey: "This lobby only supports the v1 pull flow."])
        }
        let wire = try PeerOfferStore.shared.prepareOutgoing(files: files)
        do {
            guard let url = URL(string: "http://\(lobby.host):\(lobby.peerPort)\(PeerProtocol.offerPathV2)") else {
                throw NSError(domain: "FTPortal.PeerOffer", code: 3, userInfo: [NSLocalizedDescriptionKey: "Invalid peer URL"])
            }
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.timeoutInterval = 5
            request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
            request.setValue("application/json", forHTTPHeaderField: "Accept")
            request.setValue(PeerProtocol.versionV2, forHTTPHeaderField: "X-FTPortal-Client")
            request.httpBody = try JSONEncoder().encode(wire)

            let (_, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
                let status = (response as? HTTPURLResponse)?.statusCode ?? 0
                throw NSError(domain: "FTPortal.PeerOffer", code: 4, userInfo: [NSLocalizedDescriptionKey: "Peer rejected offer with HTTP \(status)"])
            }
            return wire
        } catch {
            PeerOfferStore.shared.cancelOutgoing(wire.offerId)
            throw error
        }
    }
}

enum PeerOfferReceiver {
    static func receiveAll(_ offer: IncomingPeerOffer) async -> PeerOfferReceiveResult {
        let directory = FileManager.default.temporaryDirectory
            .appendingPathComponent("FTPortal-Offer-\(UUID().uuidString)", isDirectory: true)
        do {
            try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        } catch {
            return PeerOfferReceiveResult(urls: [], directory: nil, failures: [error.localizedDescription])
        }

        var urls: [URL] = []
        var failures: [String] = []
        for file in offer.files {
            do {
                guard let url = URL(string: "http://\(offer.host):\(offer.peerPort)\(PeerProtocol.transferPrefixV2)\(offer.offerId)/\(file.id)") else {
                    throw NSError(domain: "FTPortal.PeerOffer", code: 5, userInfo: [NSLocalizedDescriptionKey: "Invalid URL for \(file.name)"])
                }
                var request = URLRequest(url: url)
                request.timeoutInterval = 3600
                request.setValue("Bearer \(offer.token)", forHTTPHeaderField: "Authorization")
                request.setValue(PeerProtocol.versionV2, forHTTPHeaderField: "X-FTPortal-Client")
                let (temporary, response) = try await URLSession.shared.download(for: request)
                guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                    throw NSError(domain: "FTPortal.PeerOffer", code: 6, userInfo: [NSLocalizedDescriptionKey: "\(file.name): peer rejected the transfer"])
                }

                let destination = uniqueDestination(directory: directory, name: safeFileName(file.name))
                try FileManager.default.moveItem(at: temporary, to: destination)
                if file.size >= 0 {
                    let attributes = try FileManager.default.attributesOfItem(atPath: destination.path)
                    let received = (attributes[.size] as? NSNumber)?.int64Value ?? -1
                    guard received == file.size else {
                        try? FileManager.default.removeItem(at: destination)
                        throw NSError(domain: "FTPortal.PeerOffer", code: 7, userInfo: [NSLocalizedDescriptionKey: "\(file.name): expected \(file.size) bytes, received \(received)"])
                    }
                }
                urls.append(destination)
                PeerOfferStore.shared.markIncomingFileComplete(offerId: offer.offerId, fileId: file.id)
            } catch {
                failures.append(error.localizedDescription)
            }
        }

        if urls.isEmpty { try? FileManager.default.removeItem(at: directory) }
        return PeerOfferReceiveResult(urls: urls, directory: urls.isEmpty ? nil : directory, failures: failures)
    }

    private static func safeFileName(_ name: String) -> String {
        let leaf = (name as NSString).lastPathComponent
            .replacingOccurrences(of: "\r", with: "_")
            .replacingOccurrences(of: "\n", with: "_")
        return leaf.isEmpty ? "FTPortal-download" : leaf
    }

    private static func uniqueDestination(directory: URL, name: String) -> URL {
        var candidate = directory.appendingPathComponent(name)
        if !FileManager.default.fileExists(atPath: candidate.path) { return candidate }
        let source = name as NSString
        let stem = source.deletingPathExtension
        let ext = source.pathExtension
        var index = 2
        while true {
            let suffix = ext.isEmpty ? "\(stem) (\(index))" : "\(stem) (\(index)).\(ext)"
            candidate = directory.appendingPathComponent(suffix)
            if !FileManager.default.fileExists(atPath: candidate.path) { return candidate }
            index += 1
        }
    }
}

File: WINDOWS/PeerOffers.cs
using System.Collections.Concurrent;
using System.Net;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace FTPortal.Windows;

internal sealed record PeerOfferFile(string Id, string Name, long Size, string Mime);

internal sealed record PeerOfferWire(
    [property: JsonPropertyName("protocol")] string Protocol,
    string OfferId,
    string SenderDeviceId,
    string SenderAlias,
    string SenderPlatform,
    int PeerPort,
    string Token,
    string VerificationCode,
    long ExpiresAt,
    IReadOnlyList<PeerOfferFile> Files
);

internal sealed record IncomingPeerOffer(
    string OfferId,
    string SenderDeviceId,
    string SenderAlias,
    string SenderPlatform,
    string Host,
    int PeerPort,
    string Token,
    string VerificationCode,
    long ExpiresAt,
    IReadOnlyList<PeerOfferFile> Files
);

internal sealed record PeerOfferReceiveResult(IReadOnlyList<string> Completed, IReadOnlyList<string> Failures);

internal sealed class OutgoingPeerGrant(string token, long expiresAt, IEnumerable<string> fileIds)
{
    public string Token { get; } = token;
    public long ExpiresAt { get; } = expiresAt;
    public HashSet<string> FileIds { get; } = new(fileIds, StringComparer.OrdinalIgnoreCase);
}

internal static class PeerOfferStore
{
    private const long OfferTtlMs = 5 * 60 * 1000L;
    private const long MaximumClockWindowMs = 10 * 60 * 1000L;
    private const int MaximumIncomingOffers = 20;
    private const int MaximumFilesPerOffer = 128;
    private static readonly object Gate = new();
    private static readonly Dictionary<string, IncomingPeerOffer> IncomingById = new(StringComparer.OrdinalIgnoreCase);
    private static readonly Dictionary<string, OutgoingPeerGrant> OutgoingById = new(StringComparer.OrdinalIgnoreCase);
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        PropertyNameCaseInsensitive = true
    };

    public static PeerOfferWire PrepareOutgoing(IReadOnlyList<SharedFile> files)
    {
        if (files.Count == 0) throw new InvalidOperationException("No pending files to offer.");
        if (files.Count > MaximumFilesPerOffer) throw new InvalidOperationException("Too many files in one offer.");
        if (files.Any(file => file.Size < 0)) throw new InvalidOperationException("All v2 offer files must have a known size.");

        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var offerId = Convert.ToHexString(RandomNumberGenerator.GetBytes(16)).ToLowerInvariant();
        var token = Convert.ToHexString(RandomNumberGenerator.GetBytes(32)).ToLowerInvariant();
        var code = RandomNumberGenerator.GetInt32(0, 1_000_000).ToString("D6");
        var expiresAt = now + OfferTtlMs;
        var manifest = files.Select(file => new PeerOfferFile(file.Id, file.Name, file.Size, file.Mime)).ToArray();
        var wire = new PeerOfferWire(
            PeerProtocol.VersionV2,
            offerId,
            PeerIdentity.DeviceId(),
            PeerIdentity.Alias(),
            "Windows",
            PeerProtocol.Port,
            token,
            code,
            expiresAt,
            manifest
        );

        lock (Gate)
        {
            CleanupLocked(now);
            OutgoingById[offerId] = new OutgoingPeerGrant(token, expiresAt, manifest.Select(file => file.Id));
        }
        return wire;
    }

    public static void CancelOutgoing(string offerId)
    {
        lock (Gate) OutgoingById.Remove(offerId);
    }

    public static IncomingPeerOffer? ReceiveIncoming(string remoteHost, string json)
    {
        PeerOfferWire? wire;
        try { wire = JsonSerializer.Deserialize<PeerOfferWire>(json, JsonOptions); }
        catch (JsonException) { return null; }
        if (wire is null) return null;

        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        if (!string.Equals(wire.Protocol, PeerProtocol.VersionV2, StringComparison.Ordinal) ||
            wire.Files is null ||
            string.IsNullOrWhiteSpace(wire.Token) ||
            string.IsNullOrWhiteSpace(wire.OfferId) ||
            string.IsNullOrWhiteSpace(wire.VerificationCode)) return null;
        if (!IsHex(wire.OfferId, 32) || !IsHex(wire.Token, 64)) return null;
        if (string.IsNullOrWhiteSpace(wire.SenderDeviceId) || string.Equals(wire.SenderDeviceId, PeerIdentity.DeviceId(), StringComparison.OrdinalIgnoreCase)) return null;
        if (wire.VerificationCode.Length != 6 || !wire.VerificationCode.All(char.IsDigit)) return null;
        if (wire.ExpiresAt <= now || wire.ExpiresAt > now + MaximumClockWindowMs) return null;
        if (wire.PeerPort is < 1 or > 65535 || string.IsNullOrWhiteSpace(remoteHost)) return null;
        if (wire.Files.Count is < 1 or > MaximumFilesPerOffer) return null;

        var files = wire.Files
            .Where(file => file is not null && !string.IsNullOrWhiteSpace(file.Id) && file.Id.Length <= 64 && file.Id.All(char.IsLetterOrDigit) && file.Size >= 0)
            .Where(file => !string.IsNullOrWhiteSpace(file.Name) && file.Name.Length <= 255)
            .Select(file => file with
            {
                Name = file.Name,
                Mime = string.IsNullOrWhiteSpace(file.Mime) ? "application/octet-stream" : file.Mime[..Math.Min(file.Mime.Length, 128)]
            })
            .ToArray();
        if (files.Length != wire.Files.Count || files.Select(file => file.Id).Distinct(StringComparer.OrdinalIgnoreCase).Count() != files.Length) return null;

        var alias = string.IsNullOrWhiteSpace(wire.SenderAlias) ? remoteHost : wire.SenderAlias[..Math.Min(wire.SenderAlias.Length, 120)];
        var platform = string.IsNullOrWhiteSpace(wire.SenderPlatform) ? "Unknown" : wire.SenderPlatform[..Math.Min(wire.SenderPlatform.Length, 40)];
        var offer = new IncomingPeerOffer(
            wire.OfferId.ToLowerInvariant(),
            wire.SenderDeviceId,
            alias,
            platform,
            remoteHost,
            wire.PeerPort,
            wire.Token.ToLowerInvariant(),
            wire.VerificationCode,
            wire.ExpiresAt,
            files
        );

        lock (Gate)
        {
            CleanupLocked(now);
            if (!IncomingById.ContainsKey(offer.OfferId) && IncomingById.Count >= MaximumIncomingOffers) return null;
            IncomingById[offer.OfferId] = offer;
        }
        return offer;
    }

    public static IReadOnlyList<IncomingPeerOffer> Incoming()
    {
        lock (Gate)
        {
            CleanupLocked(DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
            return IncomingById.Values.OrderBy(offer => offer.SenderAlias, StringComparer.OrdinalIgnoreCase).ToArray();
        }
    }

    public static void Decline(string offerId)
    {
        lock (Gate) IncomingById.Remove(offerId);
    }

    public static void MarkIncomingFileComplete(string offerId, string fileId)
    {
        lock (Gate)
        {
            if (!IncomingById.TryGetValue(offerId, out var offer)) return;
            var remaining = offer.Files.Where(file => !string.Equals(file.Id, fileId, StringComparison.OrdinalIgnoreCase)).ToArray();
            if (remaining.Length == 0) IncomingById.Remove(offerId);
            else IncomingById[offerId] = offer with { Files = remaining };
        }
    }

    public static bool AuthorizeOutgoing(string offerId, string fileId, string? authorization)
    {
        OutgoingPeerGrant? grant;
        lock (Gate)
        {
            CleanupLocked(DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
            if (!OutgoingById.TryGetValue(offerId, out grant) || !grant.FileIds.Contains(fileId)) return false;
        }
        if (string.IsNullOrWhiteSpace(authorization) || !authorization.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase)) return false;
        var supplied = authorization[7..].Trim().ToLowerInvariant();
        var expectedBytes = Encoding.ASCII.GetBytes(grant.Token);
        var suppliedBytes = Encoding.ASCII.GetBytes(supplied);
        return expectedBytes.Length == suppliedBytes.Length && CryptographicOperations.FixedTimeEquals(expectedBytes, suppliedBytes);
    }

    public static void CompleteOutgoingFile(string offerId, string fileId)
    {
        lock (Gate)
        {
            if (!OutgoingById.TryGetValue(offerId, out var grant)) return;
            grant.FileIds.Remove(fileId);
            if (grant.FileIds.Count == 0) OutgoingById.Remove(offerId);
        }
    }

    private static void CleanupLocked(long now)
    {
        foreach (var id in IncomingById.Where(pair => pair.Value.ExpiresAt <= now).Select(pair => pair.Key).ToArray())
            IncomingById.Remove(id);
        foreach (var id in OutgoingById.Where(pair => pair.Value.ExpiresAt <= now || pair.Value.FileIds.Count == 0).Select(pair => pair.Key).ToArray())
            OutgoingById.Remove(id);
    }

    private static bool IsHex(string value, int length) =>
        value.Length == length && value.All(Uri.IsHexDigit);
}

internal static class PeerOfferClient
{
    private static readonly HttpClient Client = new(new SocketsHttpHandler
    {
        ConnectTimeout = TimeSpan.FromSeconds(3),
        UseProxy = false
    })
    {
        Timeout = TimeSpan.FromSeconds(5)
    };

    public static async Task<PeerOfferWire> SendAsync(PeerLobby lobby, IReadOnlyList<SharedFile> files, CancellationToken cancellationToken = default)
    {
        if (!lobby.SupportsOffers) throw new InvalidOperationException("This lobby only supports the v1 pull flow.");
        var wire = PeerOfferStore.PrepareOutgoing(files);
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Post, $"http://{lobby.Host}:{lobby.PeerPort}{PeerProtocol.OfferPathV2}");
            request.Headers.TryAddWithoutValidation("X-FTPortal-Client", PeerProtocol.VersionV2);
            request.Headers.Accept.ParseAdd("application/json");
            request.Content = new StringContent(JsonSerializer.Serialize(wire, new JsonSerializerOptions(JsonSerializerDefaults.Web)), Encoding.UTF8, "application/json");
            using var response = await Client.SendAsync(request, cancellationToken);
            if (!response.IsSuccessStatusCode)
                throw new IOException($"Peer rejected offer with HTTP {(int)response.StatusCode} {response.ReasonPhrase}");
            return wire;
        }
        catch
        {
            PeerOfferStore.CancelOutgoing(wire.OfferId);
            throw;
        }
    }
}

internal static class PeerOfferReceiver
{
    private static readonly HttpClient Client = new(new SocketsHttpHandler { UseProxy = false })
    {
        Timeout = Timeout.InfiniteTimeSpan
    };

    public static async Task<PeerOfferReceiveResult> ReceiveAllAsync(
        IncomingPeerOffer offer,
        string destinationDirectory,
        TransferCenter transfers,
        CancellationToken cancellationToken = default)
    {
        var completed = new List<string>();
        var failures = new List<string>();

        try
        {
            Directory.CreateDirectory(destinationDirectory);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException or ArgumentException or NotSupportedException)
        {
            failures.Add($"Destination unavailable: {TransferError(ex)}");
            return new PeerOfferReceiveResult(completed, failures);
        }

        foreach (var file in offer.Files)
        {
            var target = UniquePath(destinationDirectory, SafeFileName(file.Name));
            var temp = target + $".ftportal-{Guid.NewGuid():N}.part";
            var transferId = transfers.Begin(
                TransferDirection.Receive,
                file.Name,
                $"{offer.SenderAlias} ({offer.Host})",
                file.Size
            );
            using var transferCancellation = CancellationTokenSource.CreateLinkedTokenSource(
                cancellationToken,
                transfers.Token(transferId)
            );
            var transferToken = transferCancellation.Token;
            var received = 0L;
            try
            {
                using var request = new HttpRequestMessage(
                    HttpMethod.Get,
                    $"http://{offer.Host}:{offer.PeerPort}{PeerProtocol.TransferPrefixV2}{offer.OfferId}/{file.Id}"
                );
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", offer.Token);
                request.Headers.TryAddWithoutValidation("X-FTPortal-Client", PeerProtocol.VersionV2);
                using var response = await Client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, transferToken);
                if (response.StatusCode != HttpStatusCode.OK)
                    throw new IOException($"{file.Name}: peer returned HTTP {(int)response.StatusCode} {response.ReasonPhrase}");

                await using (var source = await response.Content.ReadAsStreamAsync(transferToken))
                await using (var destination = new FileStream(temp, FileMode.CreateNew, FileAccess.Write, FileShare.None, 128 * 1024, useAsync: true))
                {
                    var buffer = new byte[128 * 1024];
                    while (true)
                    {
                        var read = await source.ReadAsync(buffer.AsMemory(0, buffer.Length), transferToken);
                        if (read == 0) break;
                        await destination.WriteAsync(buffer.AsMemory(0, read), transferToken);
                        received += read;
                        transfers.Update(transferId, received);
                    }
                    await destination.FlushAsync(transferToken);
                }

                transferToken.ThrowIfCancellationRequested();

                if (file.Size >= 0 && received != file.Size)
                    throw new IOException($"{file.Name}: expected {file.Size:N0} bytes, received {received:N0}");

                File.Move(temp, target);
                completed.Add(target);
                PeerOfferStore.MarkIncomingFileComplete(offer.OfferId, file.Id);
                transfers.Finish(transferId, success: true, finalBytes: received);
            }
            catch (Exception ex) when (ex is IOException or HttpRequestException or UnauthorizedAccessException or ArgumentException or NotSupportedException or OperationCanceledException)
            {
                try { if (File.Exists(temp)) File.Delete(temp); } catch { }
                transfers.Finish(transferId, success: false, detail: TransferError(ex), finalBytes: received);
                failures.Add($"{file.Name}: {TransferError(ex)}");
                if (transferToken.IsCancellationRequested || cancellationToken.IsCancellationRequested) break;
            }
        }

        return new PeerOfferReceiveResult(completed, failures);
    }

    public static Task<PeerOfferReceiveResult> ReceiveAllAsync(
        IncomingPeerOffer offer,
        string destinationDirectory,
        CancellationToken cancellationToken = default) =>
        ReceiveAllAsync(offer, destinationDirectory, new TransferCenter(), cancellationToken);

    private static string TransferError(Exception error) => error switch
    {
        OperationCanceledException => "Transfer cancelled",
        HttpRequestException => "Peer connection failed",
        UnauthorizedAccessException => "Destination access denied",
        FileNotFoundException or DirectoryNotFoundException => "Destination unavailable",
        IOException => "Transfer I/O error",
        _ => "Transfer failed"
    };

    private static string SafeFileName(string name)
    {
        var leaf = Path.GetFileName(name);
        if (string.IsNullOrWhiteSpace(leaf)) leaf = "FTPortal-download";
        foreach (var invalid in Path.GetInvalidFileNameChars()) leaf = leaf.Replace(invalid, '_');
        return leaf;
    }

    private static string UniquePath(string directory, string fileName)
    {
        var candidate = Path.Combine(directory, fileName);
        if (!File.Exists(candidate)) return candidate;
        var stem = Path.GetFileNameWithoutExtension(fileName);
        var extension = Path.GetExtension(fileName);
        for (var index = 2; ; index++)
        {
            candidate = Path.Combine(directory, $"{stem} ({index}){extension}");
            if (!File.Exists(candidate)) return candidate;
        }
    }
}
